packages/controllers/src/permissions/README.md

```diff
@@ -0,0 +1,187 @@
+// types and restricted methods
+
+const caveatSpecifications = {
+  filterArrayResponse: {
+    type: 'filterArrayResponse',
+    // If a permission has any caveats, its corresponding restricted method
+    // implementation is decorated / wrapped with the implementations of its
+    // caveats, using the caveat's decorator function.
+    decorator: ((
+        // Restricted methods and other caveats can be async, so we have to
+        // assume that the method is async.
+        method: AsyncRestrictedMethod<Json, string[]>,
+        caveat: { type: 'filterArrayResponse'; value: string[] },
+      ) =>
+      async (args: RestrictedMethodOptions<RestrictedMethodParameters>) => {
+        const result = await method(args);
+        return Array.isArray(result)
+          ? result.filter(caveat.value.includes)
+          : result;
+      }) as CaveatDecorator<Json>,
+  },
+};
+
+const permissionSpecifications = {
+  wallet_getSecretArray: {
+    // i.e. the restricted method name
+    target: 'wallet_getSecretArray',
+    methodImplementation: (
+      _args: RestrictedMethodOptions<RestrictedMethodParameters>,
+    ) => {
+      return ['secret1', 'secret2', 'secret3'];
+    },
+  },
+};
+
+const permissionController = new PermissionController({
+  caveatSpecifications,
+  messenger: controllerMessenger, // assume this was given
+  permissionSpecifications,
+  unrestrictedMethods: ['wallet_unrestrictedMethod'],
+});
+```
+
+### Adding the Permission Middleware
+
+```typescript
+// This should take place where a middleware stack is created for a particular
+// subject.
+
+// The subject could be a port, stream, socket, etc.
+const origin = getOrigin(subject);
+
+const engine = new JsonRpcEngine();
+engine.push(/* your various middleware*/);
+engine.push(permissionController.createPermissionMiddleware({ origin }));
+// Your middleware stack is now permissioned
+engine.push(/* your other various middleware*/);
+```
+
+### Calling a Restricted Method Internally
+
+```typescript
+// Sometimes, we need to call a restricted method internally, as a particular subject.
+permissionController.executeRestrictedMethod(origin, 'wallet_getSecretArray');
+
+// If the restricted method has any parameters, they are given as the third
+// argument to executeRestrictedMethod().
+permissionController.executeRestrictedMethod(origin, 'wallet_getSecret', {
+  secretType: 'array',
+});
+```
+
+### Requesting and Getting Permissions
+
+```typescript
+// From the perspective of subjects, requesting and getting permissions
+// works the same as it does with `rpc-cap`.
+const approvedPermissions = await ethereum.request({
+  method: 'wallet_requestPermissions',
+  params: [{
+    wallet_getSecretArray: {},
+  }]
+})
+
+const existingPermissions = await ethereum.request({
+  method: 'wallet_getPermissions',
+)
+```
+
+### Caveat Decorators
+
+```typescript
+// Validation / passthrough
+export function onlyArrayParams(
+  next: MethodImplementation,
```

Comment:
This example seems outdated :thinking: The other example decorator has the arguments `method` and `caveat`, with different types as well.

Also, this section could use a description. e.g. "Here are a few more example caveats".